### ISSUE TITLE
Usability in utop: must manually load various stuff before using the ppx

### ISSUE THREAD

**Issue by Armael** [janestreet/ppx_sexp_conv#1]
Hi,

I think this is more a kind of "usability bug": in order to make the following declaration work in utop "`type t = int * int [@@deriving sexp];;`", I figured out by trial and error that I had to first, in this order:

```
#require "ppx_deriving";;
#require "sexplib";;
#require "ppx_sexp_conv";;
open Sexplib.Conv;;
```

The `open Sexplib.Conv` bit being the most puzzling imo (without it it complains that `int_of_sexp` is unbound). Could this be shortened/documented a bit? I my head, a `#require "ppx_deriving";; #require "ppx_sexp_conv";;` should be sufficient...?

Cheers,
Armaël

**Comment by diml** [comment]
Note: you shouldn't have to require `"ppx_deriving"` manually, can you confirm it works without?

ppx\_sexp\_conv can probably require sexplib. I'm not sure what to do about the `open Sexplib.Conv` though, as something the basic converters will come from somewhere else. For instance if you are using core then `open Core.Std` will bring into scope all the standard converters.

**Comment by Armael** [comment]
Without requiring `ppx_deriving`, I get:

```
# #require "sexplib";;
# #require "ppx_sexp_conv";;
# open Sexplib.Conv;;
# type t = int * int [@@deriving sexp];; 
Cannot locate deriver sexp
```

**Comment by diml** [comment]
Not sure what's happening. @whitequark, any idea why we need to load ppx\_deriving first?

**Comment by whitequark** [comment]
Try `utop -dsource`:
```
# #require "ppx_sexp_conv";;
                                                                                                                                                              [@@@findlib.ppxopt ("ppx_deriving", "/home/whitequark/.opam/4.02.3/lib/ppx_core/./ppx_core.cma")];;                                                                                                                                                                                                                         
[@@@findlib.ppxopt ("ppx_deriving", "/home/whitequark/.opam/4.02.3/lib/ppx_optcomp/./ppx_optcomp.cma")];;

[@@@findlib.ppxopt ("ppx_deriving", "/home/whitequark/.opam/4.02.3/lib/ppx_driver/./ppx_driver.cma")];;

[@@@findlib.ppxopt ("ppx_deriving", "/home/whitequark/.opam/4.02.3/lib/ppx_sexp_conv/./ppx_sexp_conv_expander.cma")];;

[@@@findlib.ppxopt ("ppx_deriving", "/home/whitequark/.opam/4.02.3/lib/ppx_type_conv/./ppx_type_conv.cma")];;

[@@@findlib.ppxopt ("ppx_deriving", "/home/whitequark/.opam/4.02.3/lib/ppx_type_conv/./ppx_type_conv_deriving.cma")];;

[%%ocaml.error
  let _ =
    "Cannot load /home/whitequark/.opam/4.02.3/lib/ppx_sexp_conv/./ppx_sexp_conv.cmxs: error loading shared library: /home/whitequark/.opam/4.02.3/lib/ppx_sexp_conv/./ppx_sexp_conv.cmxs: undefined symbol: camlPpx_sexp_conv_expander"
  let _ = ""];;

#require  "ppx_sexp_conv";;
```

**Comment by whitequark** [comment]
Now if you `#require "ppx_deriving"` first:
```
# #require "ppx_deriving";;
#require  "ppx_deriving";;
# #require "ppx_sexp_conv";;
;;
;;
;;
;;
;;
;;
#require  "ppx_sexp_conv";;
```

**Comment by whitequark** [comment]
@diml oh, I see. Note how e.g. ppx_core supplies the `ppxopt(-ppx_driver) = "ppx_deriving,./whatever.cma"` option, but does not require ppx_deriving. This means that ppx_deriving might well not be loaded until right before ppx_sexp_conv itself is loaded, and then the `ppxopt` is lost.

**Comment by diml** [comment]
I see, my bad. I'll add the dependency.

Actually that's a problem I had while creating the packages and wanted to come back to: ppx_core (and a few other packages) are general purpose libraries that are not related to ppx_deriving or ppx_type_conv, so they should have a META file similar to any library. However there is currently no way to specify a dependency from a ppx\_deriving plugin to a _normal_ library. For instance I can't write a ppx\_deriving plugin that would use pcre.

**Comment by whitequark** [comment]
That's a very good point. I think ppx_deriving should use findlib to discover the package archives. I think I used to link findlib before? I don't remember what made me switch to pure-dynlink.

**Comment by diml** [comment]
Yeah I think it would be easier if ppx_deriving was using findlib

**Comment by agarwal** [comment]
This is a still a bit frustrating. You have to load ppx_deriving first. If you load ppx_sexp_conv first, there's no way out except to restart utop.

If you mistakenly 
```
utop # #require "ppx_sexp_conv";;
utop # type t = int [@@deriving sexp];;
Cannot locate deriver sexp

(* Now you google for this error and end up here. Now you realize you need to load ppx_deriving. So you try it. *)

#require "ppx_deriving";;
utop # type t = int [@@deriving sexp];;
Cannot locate deriver sexp

utop # #require "ppx_deriving";;
utop # #require "ppx_sexp_conv";;
utop # type t = int [@@deriving sexp];;
Cannot locate deriver sexp
```

Finally, restart utop and do the last sequence, and then it works.

**Comment by whitequark** [comment]
Yeah, unfortunately I haven't had time yet to fix that.

**Comment by diml** [comment]
This is fixed with ppx_sexp_conv 113.24.00

**Comment by Armael** [comment]
Ok, thanks!
